### ISSUE TITLE
Rule proposal: `no-untyped-empty-array`

### ISSUE THREAD

**Issue by zanminkian** [typescript-eslint/typescript-eslint#9493]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Empty array in ts is not type safe.
```ts
const arr = []; // type of arr is `any[]`
```

### Fail Cases

```TypeScript
const arr = [];
```

### Pass Cases

```TypeScript
const arr: number[] = [];
const arr: any[] = [];
const arr = [] as unknow[];
const arr = ['foo'];
```

**Comment by auvred** [comment]
Btw, this result can be achieved using only the [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule.

The following ESQuery selector will only match empty untyped arrays in variable declarations:
```
VariableDeclarator:not([id.typeAnnotation]) > ArrayExpression.init[elements.length=0]
```

[playground link](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&esQuery=N4IgZglgNgLgpgJxALhCAvkA&code=MYewdgzgLgBAhgJwQRgFw2gglmA5gbQF0YBeGIgKFElkQQCZTzi4IMps9Lrp4kBmJvgDkAMxAhhhKuF50ALEMJA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1JFo0fNEgAaTOGw5kiJDWHdIANTaU2AIyQARRFXiq2pYWgKkAFDwp0AdKSYAHRAEFcpwxXy4AugEowAPjAnQTYmAFEOexJkZA9cawpcClIeWUQAW0RWZGskXABzUgALAF4ABi8xSSlITJi2fMRFLU8AclIwWBlO1gd6MFVoUNRqsABfL0kxkDGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by bradzacher** [comment]
I'm a -1 for the same reasons - it's trivial to do VIA `no-restricted-syntax` and the arrays aren't actually untyped because TS almost always can infer a type from the `.push` calls.

Three maintainers against and some good reasons - were going to reject this proposal.
